### ISSUE TITLE
File Not Found org.apache.http.impl.execchain.RetryExec execute INFO: I/O exception (No such file or directory)

### ISSUE THREAD

**Issue by LeYves** [maximilianvoss/spotify-toniebox-sync#39]
Hi, I have managed to configure and execute musicsync on RaspberryPI. It seems like the playlist is being read from Spotify but then I get an error when the file is to be stored in the cache directory.
I have tried multiple variations of the path in the config e.g. absolute path and so on. I also set permissions to 777. In the output below the cache dir. is defined in the config to: "cachePath": "file_cache3" . The application created the file it self under  ~/musicsync/package/application/target/ after startup, so I am assuming that the application has all rights required to create directories and files.

I also installed FFmpeg according to this issue: https://github.com/maximilianvoss/musicsync/issues/29. This issue states also a lack of error output. Is there a fix ?

Here is my Output: ( So far I have concidered the warnings about multiple bindings of SLF4J as irrelevant )
Thanks for any help !

pi@raspberrypi:~/musicsync/package/application/target $ musicsync
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:///home/pi/musicsync/package/application/target/./modules/log4j-slf4j-impl-2.14.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/pi/musicsync/package/application/target/modules/log4j-slf4j-impl-2.14.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.apache.logging.slf4j.Log4jLoggerFactory]
2021-12-24 20:50:31,888 main WARN Could not locate file file:/home/pi/musicsync/package/application/target/log4j2.xml
2021-12-24 20:50:31,899 main WARN Could not locate file file:/home/pi/musicsync/package/application/target/log4j2.xml
Dec 24, 2021 8:50:50 PM org.apache.http.impl.execchain.RetryExec execute
INFO: I/O exception (java.io.FileNotFoundException) caught when processing request to {s}->https://bxn-toniecloud-prod-upload.s3.amazonaws.com:443: file_cache3/6KxcW9t7ae5kPntqfYtKtv.mp3 (No such file or directory)
Dec 24, 2021 8:50:50 PM org.apache.http.impl.execchain.RetryExec execute
INFO: Retrying request to {s}->https://bxn-toniecloud-prod-upload.s3.amazonaws.com:443
Dec 24, 2021 8:50:51 PM org.apache.http.impl.execchain.RetryExec execute
INFO: I/O exception (java.io.FileNotFoundException) caught when processing request to {s}->https://bxn-toniecloud-prod-upload.s3.amazonaws.com:443: file_cache3/6KxcW9t7ae5kPntqfYtKtv.mp3 (No such file or directory)
Dec 24, 2021 8:50:51 PM org.apache.http.impl.execchain.RetryExec execute
INFO: Retrying request to {s}->https://bxn-toniecloud-prod-upload.s3.amazonaws.com:443
Dec 24, 2021 8:50:51 PM org.apache.http.impl.execchain.RetryExec execute
INFO: I/O exception (java.io.FileNotFoundException) caught when processing request to {s}->https://bxn-toniecloud-prod-upload.s3.amazonaws.com:443: file_cache3/6KxcW9t7ae5kPntqfYtKtv.mp3 (No such file or directory)
Dec 24, 2021 8:50:51 PM org.apache.http.impl.execchain.RetryExec execute
INFO: Retrying request to {s}->https://bxn-toniecloud-prod-upload.s3.amazonaws.com:443
20:50:51.888 [main] ERROR rocks.voss.musicsync.plugins.output.toniebox.TonieboxOutputPlugin - Exception
java.io.FileNotFoundException: file_cache3/6KxcW9t7ae5kPntqfYtKtv.mp3 (No such file or directory)
	at java.io.FileInputStream.open0(Native Method) ~[?:?]
	at java.io.FileInputStream.open(FileInputStream.java:219) ~[?:?]
	at java.io.FileInputStream.<init>(FileInputStream.java:157) ~[?:?]
	at org.apache.http.entity.mime.content.FileBody.writeTo(FileBody.java:116) ~[httpmime-4.5.13.jar:4.5.13]
	at org.apache.http.entity.mime.AbstractMultipartForm.doWriteTo(AbstractMultipartForm.java:134) ~[httpmime-4.5.13.jar:4.5.13]
	at org.apache.http.entity.mime.AbstractMultipartForm.writeTo(AbstractMultipartForm.java:157) ~[httpmime-4.5.13.jar:4.5.13]
	at org.apache.http.entity.mime.MultipartFormEntity.writeTo(MultipartFormEntity.java:113) ~[httpmime-4.5.13.jar:4.5.13]
	at org.apache.http.impl.DefaultBHttpClientConnection.sendRequestEntity(DefaultBHttpClientConnection.java:156) ~[httpcore-4.4.13.jar:4.4.13]
	at org.apache.http.impl.conn.CPoolProxy.sendRequestEntity(CPoolProxy.java:152) ~[httpclient-4.5.13.jar:4.5.13]
	at org.apache.http.protocol.HttpRequestExecutor.doSendRequest(HttpRequestExecutor.java:238) ~[httpcore-4.4.13.jar:4.4.13]
	at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:123) ~[httpcore-4.4.13.jar:4.4.13]
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:272) ~[httpclient-4.5.13.jar:4.5.13]
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186) ~[httpclient-4.5.13.jar:4.5.13]
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89) ~[httpclient-4.5.13.jar:4.5.13]
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110) ~[httpclient-4.5.13.jar:4.5.13]
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185) ~[httpclient-4.5.13.jar:4.5.13]
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83) ~[httpclient-4.5.13.jar:4.5.13]
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108) ~[httpclient-4.5.13.jar:4.5.13]
	at rocks.voss.toniebox.RequestHandler.executeRequest(RequestHandler.java:183) ~[toniebox-api-3.3.jar:?]
	at rocks.voss.toniebox.RequestHandler.executePostRequest(RequestHandler.java:154) ~[toniebox-api-3.3.jar:?]
	at rocks.voss.toniebox.RequestHandler.uploadFile(RequestHandler.java:122) ~[toniebox-api-3.3.jar:?]
	at rocks.voss.toniebox.beans.toniebox.CreativeTonie.uploadFile(CreativeTonie.java:76) ~[toniebox-api-3.3.jar:?]
	at rocks.voss.musicsync.plugins.output.toniebox.TonieboxOutputPlugin.uploadTrack(TonieboxOutputPlugin.java:57) [musicsync-plugins-output-toniebox-3.0.jar:?]
	at rocks.voss.musicsync.plugins.output.toniebox.TonieboxOutputPlugin.uploadTracks(TonieboxOutputPlugin.java:41) [musicsync-plugins-output-toniebox-3.0.jar:?]
	at rocks.voss.musicsync.application.Application.syncBulk(Application.java:140) [musicsync-application-3.0.jar:?]
	at rocks.voss.musicsync.application.Application.sync(Application.java:98) [musicsync-application-3.0.jar:?]
	at rocks.voss.musicsync.application.Application.main(Application.java:72) [musicsync-application-3.0.jar:?]

**Comment by maximilianvoss** [comment]
Have you tried to define an absolute path for the caching directory?

**Comment by maximilianvoss** [comment]
As I do not see the Spotify calls, is Spotify indeed running on you host? Does the cache file exist?

**Comment by maximilianvoss** [comment]
Checked it, either it is due to a not absolute path or that stream_recording.pl created a directory due to some code flaws which are fixed now

**Comment by LeYves** [comment]
Thanks Maximilian. Problem seems to be gone. At least I do not get any Errror messages anymore. Unfortunatly the only Output I get now is:

ubuntu@ubuntu2004:~/musicsync/package/application/target$ ./musicsync 
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:///home/ubuntu/musicsync/package/application/target/./modules/log4j-slf4j-impl-2.17.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/ubuntu/musicsync/package/application/target/modules/log4j-slf4j-impl-2.17.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.apache.logging.slf4j.Log4jLoggerFactory]

And then nothing anymore. 
The Cache folder is beeing created ( I am using anabsolute path ) but I see no files inside.
I also do not get any output in musicsync.log so the only lifesign I have are the above lines and the fact that the Cache folder is being created.
And of course files are not being synced between Spotify Playlist and Creative Toni

Some History:
I did have the following Problems with log4j which I thought I had fixed more or less. But this might be the reason why I do not get any output in musicsync.log.

Right after the fresh compile I got:
_ubuntu@ubuntu2004:~/musicsync/package/application/target$ ./musicsync 
error occurred during initialization of boot layer
java.lang.module.FindException: Two versions of module org.apache.logging.log4j.core found in ./modules (log4j-core-2.17.1.jar and log4j-core-2.14.1.jar)_
So I fixed that by removing log4j-core-2.14.1.jar from ~/musicsync/package/application/target/modules

Then I got:
_ubuntu@ubuntu2004:~/musicsync/package/application/target$ ./musicsync 
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:///home/ubuntu/musicsync/package/application/target/./modules/log4j-slf4j-impl-2.17.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/ubuntu/musicsync/package/application/target/modules/log4j-slf4j-impl-2.17.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.apache.logging.slf4j.Log4jLoggerFactory]
**2022-01-04 16:46:57,787 main WARN Could not locate file file:/home/ubuntu/musicsync/package/application/target/log4j2.xml
2022-01-04 16:46:57,790 main WARN Could not locate file file:/home/ubuntu/musicsync/package/application/target/log4j2.xml**_

So I copied the log4j2.xml file from ~/musicsync/application/src/main/resources to /home/ubuntu/musicsync/package/application/target/

Which then resulted in the posted result above, with no further output and no log file output.

_ubuntu@ubuntu2004:~/musicsync/package/application/target$ ./musicsync 
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:///home/ubuntu/musicsync/package/application/target/./modules/log4j-slf4j-impl-2.17.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/ubuntu/musicsync/package/application/target/modules/log4j-slf4j-impl-2.17.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.apache.logging.slf4j.Log4jLoggerFactory]_

By the way, I switched from RaspberryPI to Ubuntu 20.04 to make shure I am not expieiencing problems with Raspbian.

Thanks for helping me move foreward. I love the idea of your project and I realy want to make it work for me. (I mean my Kids of course ;-)

Grüße Yves

**Comment by maximilianvoss** [comment]
Hi Yves,
put following log4j2.xml into your working directory
```
<?xml version="1.0" encoding="UTF-8"?>
<Configuration status="WARN">
    <Appenders>
        <Console name="Console" target="SYSTEM_OUT">
            <PatternLayout pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n"/>
        </Console>

        <File name="File" fileName="musicsync.log">
            <PatternLayout pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n"/>
        </File>
    </Appenders>
    <Loggers>
        <Root level="DEBUG">
            <AppenderRef ref="File" level="ERROR"/>
            <AppenderRef ref="Console" level="DEBUG"/>
        </Root>
    </Loggers>
</Configuration>
```

This will output full debug information to your command line and you should see that something is going on. Also ensure dbus and spotify are running. Maybe for ubuntu the docker image makes sense as it contains all necessary applications.
Best ,
Max
